framework/src/main/java/org/checkerframework/framework/flow/CFAbstractTransfer.java

```diff
@@ -366,6 +378,119 @@ public S initialStore(UnderlyingAST underlyingAST, List<LocalVariableNode> param
     return store;
   }
 
+  /**
+   * Determine whether a lambda expression is leaked.
+   *
+   * <p>This currently checks for the {@link NonLeaked} annotation, which is trusted.
+   *
+   * @param lambda the lambda expression
+   * @param aTypeFactory an annotated type factory
+   * @return true if the lambda expression is found to be annotated with {@link NonLeaked}
+   */
+  private boolean isLambdaLeaked(CFGLambda lambda, AnnotatedTypeFactory aTypeFactory) {
+    LambdaExpressionTree lambdaTree = lambda.getLambdaTree();
+    TreePath lambdaPath = aTypeFactory.getPath(lambdaTree);
+    Tree lambdaParent = lambdaPath.getParentPath().getLeaf();
+    if (lambdaParent.getKind() == Tree.Kind.METHOD_INVOCATION) {
+      MethodInvocationTree invok = (MethodInvocationTree) lambdaParent;
+      ExecutableElement methodElt = TreeUtils.elementFromUse(invok);
+      AliasingAnnotatedTypeFactory aliasingAtf =
+          analysis
+              .atypeFactory
+              .getChecker()
+              .getTypeFactoryOfSubcheckerOrNull(AliasingChecker.class);
+      if (aliasingAtf != null) {
+        // Check whether any parameter is NOT annotated with @NonLeaked
+        return methodElt.getParameters().stream()
+            .anyMatch(
+                param ->
+                    aliasingAtf.getAnnotatedType(param).getPrimaryAnnotation(NonLeaked.class)
+                        == null);
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Returns true if a given lambda expression is pure. A lambda expression is pure if:
+   *
+   * <ol>
+   *   <li>All methods invoked within the lambda expression are known to be pure.
+   *   <li>All arguments to method calls within the lambda expression are not assignable by other
+   *       code
+   * </ol>
+   *
+   * @param lambda a lambda expression
+   * @param aTypeFactory an annotated type factory
+   * @return true if the given lambda expression is pure
+   */
+  private boolean isLambdaPure(CFGLambda lambda, AnnotatedTypeFactory aTypeFactory) {
```

Comment:
Why not write an "isExpressionPure" method rather than an "isLambdaPure" method?
Even better, reuse code from the Purity Checker rather than re-implementing it partially here.  For example, could you do `h = new PurityCheckerHelper(...); h.scan( my-expression-or-statement ); h.purityResult`?  If re-use of `PurityChecker` is not possible, please document why.